src/System.Net.Http/src/System/Net/Http/Unix/CurlResponseMessage.cs

```diff
@@ -0,0 +1,247 @@
+            Content = new StreamContent(_responseStream);
+        }
+    }
+
+        internal sealed class CurlResponseStream : Stream
+        {
+            private volatile bool _disposed = false;
+            private Stream _innerStream;
+            private readonly object _lockObject = new object();
+            private readonly CancellationTokenSource _cancellationTokenSource = new CancellationTokenSource();
+            private readonly ManualResetEventSlim _dataRequestedEvent = new ManualResetEventSlim(false);
+            private readonly ManualResetEventSlim _dataReadyEvent = new ManualResetEventSlim(false);
+            private readonly ManualResetEventSlim _noMoreDataEvent = new ManualResetEventSlim(false);
+
+            internal CurlResponseStream() : base()
+            {
+            }
+
+            internal void SignalComplete(bool isCancellation)
+            {
+                if (!_disposed)
+                {
+                    _cancellationTokenSource.Cancel();
+                }
+            }
+
+            internal unsafe void WaitAndSignalReaders(byte* pointer, long length)
+            {
+                CheckDisposed();
+                _dataRequestedEvent.Wait(Timeout.InfiniteTimeSpan, _cancellationTokenSource.Token);
+                lock (_lockObject)
+                {
+                    _innerStream = new UnmanagedMemoryStream(pointer, length);
+                    _noMoreDataEvent.Reset();
+                    _dataRequestedEvent.Reset();
+                    _dataReadyEvent.Set();
+                }
+                _noMoreDataEvent.Wait(Timeout.InfiniteTimeSpan, _cancellationTokenSource.Token);;
+            }
+
+            public override bool CanRead
+            {
+                get
+                {
+                    return !_disposed;
+                }
+            }
+
+            public override bool CanSeek
+            {
+                get
+                {
+                    return false;
+                }
+            }
+
+            public override bool CanWrite
+            {
+                get
+                {
+                    return false;
+                }
+            }
+
+            public override long Length
+            {
+                get
+                {
+                    CheckDisposed();
+                    throw new NotSupportedException();
+                }
+            }
+
+            public override long Position
+            {
+                get
+                {
+                    CheckDisposed();
+                    throw new NotSupportedException();
+                }
+
+                set
+                {
+                    CheckDisposed();
+                    throw new NotSupportedException();
+                }
+            }
+
+            public override void Flush()
+            {
+                // Nothing to do.
+            }
+
+            public override int Read(byte[] buffer, int offset, int count)
```

Comment:
I understand your argument for why the writer needs to block until the data is consumed, because of how libcurl works with regards to handing out a pointer to the data that's only good for the duration of the synchronous call.  But what about the reader... could we override ReadAsync, and implement it to use asynchronous waiting when data isn't available yet? e.g. awaiting on a TaskCompletionSource instead of blocking on a ManualResetEventSlim?